src/Analyzers/CodeAnalysisResult.cs

```diff
@@ -1,11 +1,40 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
-using System;
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+using NonBlocking;
 
 namespace Microsoft.CodeAnalysis.Tools.Analyzers
 {
     public class CodeAnalysisResult
     {
-        internal void AddDiagnostic(Document document, Diagnostic diagnostic) => throw new NotImplementedException();
+        private readonly ConcurrentDictionary<DocumentId, IList<Diagnostic>> _dictionary
+            = new ConcurrentDictionary<DocumentId, IList<Diagnostic>>();
+
+        internal void AddDiagnostic(Document document, Diagnostic diagnostic)
+        {
+            _ = _dictionary.AddOrUpdate(document.Id,
+                addValueFactory: (key) =>
+                 {
+                     var list = new List<Diagnostic> { diagnostic };
+                     return list;
+                 },
+                updateValueFactory: (key, list) =>
+                {
+                    list.Add(diagnostic);
+                    return list;
+                });
+        }
+
+        public IReadOnlyDictionary<DocumentId, ImmutableArray<Diagnostic>> Diagnostics
```

Comment:
I need this to proceed with PerDocumentCodeFixApplier

Reply:
yep.  I need to fix integration tests first.  You can assume that this is the shape of the object you will be given